### PR TITLE
Restore `@babel/types` support for old TS versions

### LINKED ISSUE #12519: Version 7.12.11 does not compile
## Bug Report

**Current behavior**
I use babel in a react application. When trying to produce a build (yarn run build), it fails with error code:
TypeScript error in /var/jenkins/.../node_modules/@babel/types/lib/index.d.ts(1450,50):
'{' or ';' expected.  TS1144
```
      1448 | }): ReturnedChild[];
      1449 | 
    > 1450 | declare function assertNode(node?: any): asserts node is Node;
           |                                                  ^
      1451 | 
```
As a current fix I declared a yarn resolution to use version 7.12.10, which compiles without any issues.
```
- Babel version:  v7.12.11
- Node 10.15.3, yarn 1.22.4, typescript 3.2.4
- OS: Windows 10
- How you are using Babel: webpack
- ```

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #12519
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

This PR swaps `lib/index.d.ts` and `lib/index-ts3.7.d.ts` in `@babel/types`. By doning so, older TS version that don't support the `typesVersion` option in `package.json` (it has been introduced in 3.1, but it seems not to work correctly in 3.2) will still get a working `.d.ts` file.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/12521"></a>

### CODE REVIEW

**Comment by zxbodya** [comment]
might be a bit more light weight option to try - would it fix the problem if we just add  `types` field pointing to legacy types(I think that is supported in old versions), and `typesVersion` with proper rules for newer versions already supporting it?

**Comment by nicolo-ribaudo** [comment]
Oh that's a good idea

**Comment by zxbodya** [comment]
> Oh that's a good idea

just tried it locally with typescript 3.1 - it works 😄

**Comment by nicolo-ribaudo** [comment]
Updated to use `types`. I still renamed `index-ts3.7.d.ts` to `index-legacy.d.ts`, since the types it contains are _not_ compatible with TS 3.7.

File: packages/babel-types/package.json [hunk 1]
```diff
@@ -14,10 +14,11 @@
     "directory": "packages/babel-types"
   },
   "main": "lib/index.js",
+  "types": "lib/index-legacy.d.ts",
   "typesVersions": {
-    "<3.7": {
-      "lib/index.d.ts": [
-        "lib/index-ts3.7.d.ts"
+    ">=3.7": {
```
Comment (zxbodya) [inline packages/babel-types/package.json:19]: should this be `<`

Comment (nicolo-ribaudo) [reply packages/babel-types/package.json:19]: No: `types` points to the `.d.ts` file for old TS versions. For new TS versions, (>= 3.7), we remap it back to `index.d.ts`.

**Review by zxbodya — APPROVED** [review]

**Review by JLHwung — APPROVED** [review]

**Review by existentialism — APPROVED** [review]

**Pull request merged**